homeassistant/components/homekit/covers.py

```diff
@@ -0,0 +1,84 @@
+"""Class to hold all cover accessories."""
+import logging
+
+from homeassistant.components.cover import ATTR_CURRENT_POSITION
+from homeassistant.helpers.event import async_track_state_change
+
+from . import TYPES
+from .accessories import HomeAccessory
+from .const import (
+    SERVICES_WINDOW_COVERING, CHAR_CURRENT_POSITION,
+    CHAR_TARGET_POSITION, CHAR_POSITION_STATE)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@TYPES.register('Window')
+class Window(HomeAccessory):
+    """Generate a Window accessory for a cover entity.
+
+    The cover entity must support: set_cover_position.
+    """
+
+    def __init__(self, hass, entity_id, display_name):
+        """Initialize a Window accessory object."""
+        super().__init__(display_name)
+        self.set_category(self.ALL_CATEGORIES.WINDOW)
+        self.set_accessory_info(entity_id)
+        self.add_preload_service(SERVICES_WINDOW_COVERING)
+
+        self._hass = hass
+        self._entity_id = entity_id
+
+        self.current_position = None
+        self.homekit_target = None
+
+        self.service_cover = self.get_service(SERVICES_WINDOW_COVERING)
+        self.char_current_position = self.service_cover. \
+            get_characteristic(CHAR_CURRENT_POSITION)
+        self.char_target_position = self.service_cover. \
+            get_characteristic(CHAR_TARGET_POSITION)
+        self.char_position_state = self.service_cover. \
+            get_characteristic(CHAR_POSITION_STATE)
+
+        self.char_target_position.setter_callback = self.move_cover
+
+    def run(self):
+        """Method called be object after driver is started."""
+        state = self._hass.states.get(self._entity_id)
+        self.update_cover_position(new_state=state)
+
+        async_track_state_change(
+            self._hass, self._entity_id, self.update_cover_position)
+
+    def move_cover(self, value):
+        """Move cover to value if call came from homekit."""
+        if value != self.current_position:
```

Comment:
Is this up to us to filter out? What if a cover has an assumed state and our assumption is off?

(we can change this in a future PR if we want)

Reply:
`self.current_position` isn't the same as the state attribute `current_position`. Unfortunatly right now the way pyhap handles changes in HomeKit is by calling the same method as HomeAssistant is calling for changes to the Homekit Status. Without this if clause something like this could happen:
* HomeKit sets cover position form `0` to `100`
* HomeAssistant state listener triggers the same method, after an initial move of the cover and sets its position to the current_position attribute of the `new_state`, which almost always won't be `100` (more like `10` or so)

Their might be a better way to do it though, but thats sertainly a topic for another PR.